### PR TITLE
#1296 Port EventBusUnsubscribers

### PR DESCRIPTION
**Pull request by maxim-s**

### CODE REVIEW

File: src/core/Akka/Event/EventBusUnsubscriber.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+           
+        }
+
+        protected override bool Receive(object message)
+        {
+            return message.Match().With<Register>(register =>
+            {
+                if (_debug)
+                    _eventStream.Publish(new Debug(this.GetType().Name, GetType(),
+                        $"watching {register.Actor} in order to unsubscribe from EventStream when it terminates"));
```
Comment (Horusiath) [inline src/core/Akka/Event/EventBusUnsubscriber.cs]: We still have VS2013 users and compatibility with C# 5 / .NET 4.5. I don't think this is valid notation.

File: src/core/Akka/Event/EventBusUnsubscriber.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+        protected override void PreStart()
+        {
+            if (_debug)
+                _eventStream.Publish(new Debug(this.GetType().Name, GetType(),
+                    $"registering unsubscriber with {_eventStream}"));
+            _eventStream.InitUnsubscriber(Self, _system);
+        }
+
+        internal class Register
```
Comment (Horusiath) [inline src/core/Akka/Event/EventBusUnsubscriber.cs]: It's better to keep all messages immutable by convention. Please, don't expose mutable fields in messages.

File: src/core/Akka/Event/EventStream.cs [hunk 3]
```diff
@@ -89,9 +97,83 @@ public override bool Unsubscribe(IActorRef subscriber)
...
+                }
+            }).WasHandled;
+        }
+
+        private void RegisterWithUnsubscriber(IActorRef subscriber)
+        {
+            _initiallySubscribedOrUnsubscriber.Match().With<Left<ISet<IActorRef>>>(v =>
+            {
+                if (_initiallySubscribedOrUnsubscriber.CompareAndSet(v,
+                    Either.Left<ISet<IActorRef>>(new HashSet<IActorRef>(v.Value.Union(new[] {subscriber})))))
```
Comment (Horusiath) [inline src/core/Akka/Event/EventStream.cs]: Immutable set is better choice here, we probably don't want to rebuild a whole `HashSet` on each change. Same concerns line 167.

**Comment by rogeralsing** [comment]
Looks solid to me.
Could you also squash the commits into one? we try to minimize the commit count in PR's

(just in case: http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html)

**Comment by rogeralsing** [comment]
Great job @maxim-s !

**Pull request merged**